SFTP driver finishing

Khady opened the pull request.

Akiat committed:
SFTP driver finishing

k4nar commented on onitu/drivers/sftp/sftp.py:
```diff
@@ -0,0 +1,231 @@
+
+
+def create_dirs(sftp, path):
+    dirs = path.split('/')
```
Use `os.path.split` instead.

k4nar commented on onitu/drivers/sftp/sftp.py:
```diff
@@ -0,0 +1,231 @@
+
+    tmp_path = "./"
+    for d in dirs:
+        tmp_path += d + "/"
```
Use `os.path.join`

k4nar commented on onitu/drivers/sftp/sftp.py:
```diff
@@ -0,0 +1,231 @@
+    for d in dirs:
+        tmp_path += d + "/"
+        try:
+            sftp.stat(tmp_path)
```
You don't need to stat it if the parent was not existing.

k4nar commented on onitu/drivers/sftp/sftp.py:
```diff
@@ -0,0 +1,231 @@
+
+
+def get_full_path(filename):
+    global root
```
Unnecessary.

k4nar commented on onitu/drivers/sftp/sftp.py:
```diff
@@ -0,0 +1,231 @@
+
+def get_full_path(filename):
+    global root
+    return root + '/' + filename
```
Use `os.path.join`.

k4nar commented on onitu/drivers/sftp/sftp.py:
```diff
@@ -0,0 +1,231 @@
+        data = f.readv([(0, stats.st_size)])
+
+        for d in data:
+            return d
```
WTF?

k4nar commented on onitu/drivers/sftp/sftp.py:
```diff
@@ -0,0 +1,231 @@
+    try:
+        stats = sftp.stat(full_path)
+        f = sftp.open(full_path, 'r')
+        data = f.readv([(0, stats.st_size)])
```
I don't think that one big read is the best option here. We could try to use a chunked get, and return the result.

k4nar commented on onitu/drivers/sftp/sftp.py:
```diff
@@ -0,0 +1,231 @@
+    full_path = get_full_path(metadata.filename)
+
+    events_to_ignore.add(full_path)
+    create_dirs(sftp, str(os.path.dirname(metadata.filename)))
```
Why the `str` ?

k4nar commented on onitu/drivers/sftp/sftp.py:
```diff
@@ -0,0 +1,231 @@
+
+    try:
+        stat_res = sftp.stat(full_path)
+        metadata.size = stat_res.st_size
```
Why are you updating the size ?

k4nar commented on onitu/drivers/sftp/sftp.py:
```diff
@@ -0,0 +1,231 @@
+                metadata.filename, e)
+        )
+
+    if metadata.filename in events_to_ignore:
```
It should be in a `finally` block.

k4nar commented on onitu/drivers/sftp/sftp.py:
```diff
@@ -0,0 +1,231 @@
+    create_dirs(sftp, str(os.path.dirname(metadata.filename)))
+
+    try:
+        sftp.open(full_path, 'w+').close()
```
Do we really need that ? We could create it when uploading the content, and avoid two calls.

k4nar commented on onitu/drivers/sftp/sftp.py:
```diff
@@ -0,0 +1,231 @@
+class CheckChanges(threading.Thread):
+
+    def __init__(self, timer, sftp):
+        threading.Thread.__init__(self)
```
Use `super` instead.

k4nar commented on onitu/drivers/sftp/sftp.py:
```diff
@@ -0,0 +1,231 @@
+    return root + '/' + filename
+
+
+class CheckChanges(threading.Thread):
```
Your thread should be a daemon or you will have an infinite loop when closing.

k4nar commented on onitu/drivers/sftp/sftp.py:
```diff
@@ -0,0 +1,231 @@
+
+
+def start():
+    global plug
```
Nope.

k4nar commented on onitu/drivers/sftp/sftp.py:
```diff
@@ -0,0 +1,231 @@
+        for f in filelist:
+            filepath = os.path.join(path, f)
+            try:
+                stat_res = sftp.stat(str(filepath))
```
Why `str` ?

k4nar commented on onitu/drivers/sftp/sftp.py:
```diff
@@ -0,0 +1,231 @@
+    hostname = plug.options['hostname']
+    username = plug.options['username']
+    password = plug.options['password']
+    port = int(plug.options['port'])
```
The plug will return an int as you set the type in the Manifest. same thing for the timer.

k4nar commented on onitu/drivers/sftp/sftp.py:
```diff
@@ -0,0 +1,231 @@
+    sftp = paramiko.SFTPClient.from_transport(transport)
+
+    try:
+        sftp.chdir(root)
```
If we chdir at the beginning, we don't need `get_full_path`, right ?

k4nar replied:
My remark is still relevant.

Akiat replied:
I removed this function.

k4nar commented on onitu/drivers/sftp/sftp.py:
```diff
@@ -0,0 +1,231 @@
+
+    plug.listen()
+
+    check_changes.stop()
```
I don't think that will ever be called. But we will add a `stop` handler soon.

Akiat committed:
Fix pull request issues

k4nar commented:
@Akiat : Same remark that on #94 regarding the new architecture.

k4nar commented on onitu/drivers/sftp/sftp.py:
```diff
@@ -0,0 +1,220 @@
+plug = Plug()
+root = None
+sftp = None
+terminated = False
```
It doesn't seem to be used anywhere.

k4nar commented on onitu/drivers/sftp/sftp.py:
```diff
@@ -0,0 +1,220 @@
+def create_dirs(sftp, path):
+    parent_exists = True
+    tmp_path = '.' + os.sep
+    dirs = path.split(os.sep)
```
Actually I don't think that using `os.sep` is a good idea as on Windows it will return `\` but in Onitu and on SFTP this is a `/`.

k4nar commented:
Also, why can't we define `get_chunk` using this function: http://paramiko-docs.readthedocs.org/en/latest/api/sftp.html#paramiko.sftp_handle.SFTPHandle.read ?

k4nar commented on onitu/drivers/sftp/sftp.py:
```diff
@@ -0,0 +1,220 @@
+        plug.logger.error("{}: {}", root, e)
+
+    check_changes = CheckChanges(timer, sftp)
+    check_changes.start()
```
It should be launched as a daemon.

Akiat committed:
fix pull request issues

Akiat committed:
new architecture and tests

k4nar commented on drivers/sftp/onitu_sftp/tests/driver.py:
```diff
@@ -0,0 +1,108 @@
+
+
+class Driver(TestDriver):
+    SPEED_BUMP = 1
```
Are you sure you need a speed bump ? At least set it to 10 or the tests will be really slow.

k4nar commented on drivers/sftp/onitu_sftp/tests/driver.py:
```diff
@@ -0,0 +1,108 @@
+        except IOError as e:
+            print "{}: {}".format(root, e)
+
+        super(Driver, self).__init__('sftp',
```
It should be the first line (and you don't need to split it on several lines).

k4nar commented on drivers/sftp/onitu_sftp/tests/driver.py:
```diff
@@ -0,0 +1,108 @@
+        self.sftp.rmdir(str(self.root))
+
+    def mkdir(self, subdirs):
+        self.sftp.mkdir(subdirs)
```
It will not work if I call `rep2.mkdir('deep/dir/is/deep')`.

k4nar commented on drivers/sftp/onitu_sftp/tests/driver.py:
```diff
@@ -0,0 +1,108 @@
+                transport.connect(username=username, password=password)
+            else:
+                if private_key_error is not None:
+                    print "SFTP driver connection failed : {}".format(
```
Not Python 3 compatible. Also I guess you can raise an exception rather than printing it.

k4nar commented on drivers/sftp/onitu_sftp/tests/driver.py:
```diff
@@ -0,0 +1,108 @@
+        try:
+            sftp.chdir(root)
+        except IOError as e:
+            print "{}: {}".format(root, e)
```
cf https://github.com/onitu/onitu/pull/99/files#r19354095

k4nar commented on drivers/sftp/onitu_sftp/tests/driver.py:
```diff
@@ -0,0 +1,108 @@
+                    return
+                transport.connect(username=username, pkey=private_key)
+        except paramiko.AuthenticationException as e:
+            print "SFTP driver connection failed : {}".format(e)
```
cf https://github.com/onitu/onitu/pull/99/files#r19354095

k4nar commented on drivers/sftp/onitu_sftp/tests/driver.py:
```diff
@@ -0,0 +1,108 @@
+    """
+
+    def close(self):
+        self.sftp.rmdir(str(self.root))
```
You never declared this attribute…

Akiat committed:
fix some issues

Akiat committed:
Fix somes bugs in sftp tests

Akiat committed:
Add comment for tests

Akiat committed:
Add new comments, fix a bug and pep8

Scylardor committed:
SFTP driver: first refactor for new multi folder architecture

Scylardor committed:
SFTP driver: More refacto to make check changes thread work

We opt for a "one thread per folder" approach. Each thread
recursively checks all the subdirectories in search of regular
files.

- Add helper functions to make asking for a SFTP client easier

Scylardor committed:
Rework SFTP upload and download handlers to handle new architecture

Scylardor committed:
Implement move_file handler for SFTP driver

Scylardor committed:
SFTP: Implement remote deletion detection in check changes thread

Scylardor committed:
SFTP: create the check changes thread folder's remote directory

In the thread __init__, create the remote directory if it doesn't
exist yet.

Scylardor committed:
Adapt the SFTP test driver to new architecture

Scylardor committed:
Update SFTP Readme (and make it a Markdown file)

Scylardor commented:
Update on the SFTP driver.

I adapted it to the multi folder architecture. I adapted the tests too. It runs fine on both Python 2 & 3.

There only seems to be a small encoding problem occurring when we upload a file from Onitu to SFTP  in Python 3 : much like the bug fixed by commit 5e2227d730c43f8583b48e5c69e47f88988cb974 , it fails again [here](https://github.com/onitu/onitu/blob/develop/onitu/plug/router.py#L92) because it tries to call decode on a size that's `str` in Python 3 (after having passed by [L111](https://github.com/onitu/onitu/blob/develop/onitu/plug/router.py#L111)). The fix should be easy.

**EDIT**: I hotfixed that on develop with 02e4e82b993419da78815c6468d2c58afae3d624 , maybe we could do better later.

Waiting for review.

k4nar commented on drivers/sftp/MANIFEST.in:
```diff
@@ -0,0 +1,2 @@
+include README
```
The name is README.md now.

k4nar commented on drivers/sftp/manifest.json:
```diff
@@ -0,0 +1,46 @@
+      "name": "Timer for modification check",
+      "description": "Time in seconds between two checks for files modifications",
+      "type": "integer",
+      "default": 60
```
I think 1 minute is not long enough. I don't think someone using SFTP is looking for a fast synchronization, so we can do it something like every 5 minutes or so.

k4nar commented on drivers/sftp/onitu_sftp/sftp.py:
```diff
@@ -0,0 +1,325 @@
+@plug.handler()
+def get_file(metadata):
+    plug.logger.debug(u"Getting remote file {}".format(metadata.path))
+    sftp = get_SFTP_client_from_plug()
```
Not a big fan of opening a connection for each handler. Opening an SSH connection is very expensive (can be several seconds), so doing it at least three time for an upload is a pretty bad idea.
I think we should try to use only one with some locking for thread-safety. Regarding fault-tolerance, it really depends on how Paramiko handles it.

k4nar commented on drivers/sftp/onitu_sftp/sftp.py:
```diff
@@ -0,0 +1,325 @@
+    sftp = get_SFTP_client_from_plug()
+    try:
+        stat_res = sftp.stat(metadata.path)
+        update_metadata(metadata, stat_res.st_size, stat_res.st_mtime,
```
You should _not_ update the size at the end of an upload, this is already known and could confuse other services if the one returned by SFTP is wrong.

k4nar commented on drivers/sftp/onitu_sftp/sftp.py:
```diff
@@ -0,0 +1,325 @@
+                               .format(old_metadata.path, new_metadata.path,
+                                       ioe))
+    else:
+        raise ServiceError(u"File {} already exists !"
```
Well if it already exists you should delete it and move it then.

Scylardor replied:
Agreed, I got mixed up with Paramiko's SFTP server interface's [rename](http://docs.paramiko.org/en/latest/api/sftp.html#paramiko.sftp_si.SFTPServerInterface.rename), for which the Paramiko documentation advises to treat rename of an existing file as an error.

Scylardor committed:
SFTP: switched to one global connection for all handlers

- Removed move_file limitation for existing files

- Stop printing the passphrase at screen in an error message

- Add the close handler to properly close the global connection

Scylardor committed:
Fix PEP8 (for one blank line !)

Scylardor committed:
SFTP: don't update metadata.size at the end of an upload

- Remove all formats from debug messages (it's useless)

Scylardor committed:
Set default change timer to 5 minutes insteas of 1

- Update MANIFEST.in
